Implement basic SMB 1.0, 2.0, 3.0+ Network navigation: connect to hosts, list shares, navigate folders, create folders and files

tiirwaa opened the pull request:
**This PR adds initial support for SMB network drives. Users can now:**

- Discover SMB/Windows file-sharing devices on the local network using predefined SMB ports.
- Connect to a host via SMB (with authentication or anonymously).
- List available shares on the host.
- Navigate inside shares and browse folders.
- Create new folders and files within shares.
- Reuse SMB sessions to improve performance and reduce connection overhead.

**SMB Version Support:**

- Supported: SMB 1 (CIFS) through jcifs-ng
- Supported: SMB 2.0, 2.1, 3.0, 3.02, 3.1.1 through SMBJ

**Note:** SMBJ and jcifs-ng automatically negotiates the highest supported SMB version between the client and server.

**Additional Info:**

The LAN discovery currently scans the network using known SMB ports (139, 445, 4450). This could be improved in future updates for more comprehensive network detection.


This is the foundation for full SMB integration; future updates will include file operations and improved error handling.

https://github.com/user-attachments/assets/f3b8ea3f-dafd-4ef1-aacb-bafd88b9930b

Commit:
Add a button for SMB storage in the Home tab to access network drives

Commit:
Implement SMB host navigation

Added support to connect to an SMB host, list available shares,
and navigate through folders. Users can now browse network drives
directly from the app.

Commit:
fix(proguard): keep SMBJ and DCERPC classes to prevent R8 stripping

Added ProGuard rules to preserve SMBJ and DCERPC classes required for SMB
network drive support. This prevents missing class errors during release build
with R8 minification enabled.

Commit:
fix(proguard)

Commit:
Fix hardcoded text: use string resource for SMB storage label

Commit:
Implemented SMB path handling: added folder/file creation and session reuse with SMBConnectionManager for improved efficiency

Commit:
Comment cleanup

Raival-e commented:
This is looking so promising! I wanted to test the feature but I couldn't connect to a server, I'm not sure how the port is determined, because there is no input field for it in the dialog, this might be the reason why I couldn't connect.

Regarding the integration in the home screen, to make it easier to implement other types of cloud storages and servers, I would suggest the following:

1- In the home screen, there should be a '+' icon button in the same row of the "Storage" label, and when clicked, a drop-down menu (or a dialog) shows with all types of storages to add (for now it will just show SMB, but in the future, there will be FTP, Local Storage, etc). amd from there the dialog of adding new SMB will show.

2- The added storages should show under "Storage" section, so I'd suggest removing the separate section for the SMB from home screen to make it cleaner. 

3- the storages that are added should be saved (so that we don't need to rewrite all the credentials every time the app is opened), you can make a data class for SMB storage with all the required credentials (and other information like name, type, etc), and saved it as a set to the datastore in PreferencesManager. Then those will be retrieved and displayed in storage section as discussed in second point.

4- The design of the "add new SMB dialog" must follow the style of the other dialogs in the app. See "Jump to path" dialog for reference.

Regarding the SMB implementation, I'm not familiar with the libraries used, so I need to test the app and see if any further improvements required.

Huge thanks for your contribution, and I'm looking forward to test the app after fixing the bug regarding connecting to a server.

Commit:
Remove redundant folder check in open SMBFileHolder

Folders are already filtered in the files list, making this check unnecessary.
This function is only invoked for files, not folders.

Commit:
Remove unused toggleImportPrefsDialogSMB function

Given the current state of this PR (opening files is not implemented yet),
keeping this function could cause issues. It has been removed to prevent
unintended behavior.

Commit:
Refactor AddSMBDriveDialog to match JumpToPathDialog design

tiirwaa commented:
![bec49ffb-2049-4cda-8882-f6ba188fe96b](https://github.com/user-attachments/assets/5647bdde-0884-4f2e-8477-52cd98d8246e)
![e8d5b1d4-4575-41d3-ba25-19686f087747](https://github.com/user-attachments/assets/91b43639-3b27-41f1-9acd-f2b0c7c62772)

Commit:
Add support for custom SMB port (default 445)

Commit:
Pass parameter port in findFile, createSubFolder, createSubFile

Raival-e commented:
I couldn't make it work, it keep showing this error message. What am I missing? I used Mixplorer to create the server, and I tested it using Amaze File Explorer.

<img width="1440" height="2903" alt="Screenshot_20250907_195449_Prism File Explorer" src="https://github.com/user-attachments/assets/3a38a488-4ef3-444b-a886-fdffe16ea626" />

tiirwaa commented:
**Possible cause: SMB 1.0**

Mixplorer likely uses SMB 1.0 (CIFS) for its server feature, which is not supported by SMBJ.
SMBJ supports only SMB 2.0, 2.1, 3.0, 3.02, and 3.1.1.

To properly test your setup, try sharing a folder using:


- Samba on Linux (with SMB 2.0 or newer enabled), or

- Windows file sharing, or

- Any other server that supports SMB 2.0 or above.

Raival-e commented:
> Mixplorer likely uses SMB 1.0

okay, that's probably why it didn't connect. I thought this is straightforward like FTP, I couldn't find an app that has SMB2 or higher. Anyways, I'll try from Windows.

Commit:
Initial SMB 1.0 support via jcifs-ng: allows connecting and listing folders

Raival-e commented:
Thanks, it does connect now, however, using the same setup as before (using mixplorer to make the server), no files are shown, even though when opening the same server on MT Manager, for example, there are shared files.

Prism (no shared files shown):
https://github.com/user-attachments/assets/9408bd70-4f82-4d64-8fbc-c6bbab812b92

MT Manager:
https://github.com/user-attachments/assets/5f22790d-acf0-4a5c-b447-66eb0b85f77c

tiirwaa commented:
I tested it on my phone with MiXplorer, and it works. However, I noticed that if I enter the wrong port, username, or password, the same thing you showed in the video happens — it doesn’t detect the error. I’m going to check that. Still, I can list the shares, view the contents, and create folders and files correctly, both in SMB1 using JCIFS and in SMB2 (Windows and Linux) using SMBJ.

Commit:
Fix: Correct SMB1 connection validation; now caches contexts/sessions only if connection succeeds, preventing invalid credentials from being stored

Commit:
Add LAN scan function and dialog

- Implements a function to scan the LAN for SMB/Windows file sharing devices.
- Displays devices in the format 'hostname (IP:port)'.
- Adds a dialog to show the scanned devices and allow selection.
- Currently, scanning is based on network and predefined known SMB ports; this could be improved in the future.

tiirwaa commented:
New LAN discovery feature! Scan your local network for SMB/Windows file-sharing devices and quickly connect to them. Currently, the scan uses predefined SMB ports, but this can be improved in the future.

https://github.com/user-attachments/assets/f3b8ea3f-dafd-4ef1-aacb-bafd88b9930b
